descqa/StellarMassFunction.py

```diff
@@ -0,0 +1,353 @@
+                        '.4 < z < .5':{'zlo':.4, 'zhi':.5, 'table#':4, 'usecols':[0,11,12,13]},
+                        '.5 < z < .65':{'zlo':.5, 'zhi':.65, 'table#':4, 'usecols':[0,16,17,18]},
+                        '.65 < z < .8':{'zlo':.65, 'zhi':.8, 'table#':4, 'usecols':[0,21,22,23]},
+                        '.8 < z < 1.0':{'zlo':.8, 'zhi':1.0, 'table#':4, 'usecols':[0,26,27,28]},
+            },
+            'zrange': (0.0, 1.0),
+            'colnames': ('logM', 'log_phi', 'dlog_phi+', 'dlog_phi-'),
+            'label': 'Moustakas et. al. 2013',
+            'missingdata':'...'
+        },
+    }
+    zkey_match ='< z <'
+
+    #plotting constants
+    validation_color = 'black'
+    validation_marker = 'o'
+    default_markers = ['v', 's', 'd', 'H', '^', 'D', 'h', '<', '>', '.']
+    msize = 4 #marker-size
+
+    def __init__(self, z='redshift_true', mass='stellar_mass', Nbins=25, log_Mlo=8., log_Mhi=12.,
+                 observation='', zlo=0., zhi=1.0, zint =0.2, ncolumns=2, **kwargs):
+
+        #catalog quantities
+        self.zlabel = z
+        self.Mlabel = mass
+
+        #z-range  and mass binning
+        self.Nbins = Nbins
+        self.log_Mlo = log_Mlo
+        self.log_Mhi = log_Mhi
+        self.Mbins = np.logspace(log_Mlo, log_Mhi, Nbins+1)
+        self.DM = (log_Mhi - log_Mlo)/Nbins
+
+        #validation data
+        self.validation_data = {}
+        self.observation = observation
+
+        #check for valid observations
+        if not observation:
+            print('Warning: no data file supplied, no observation requested; only catalog data will be shown')
+        elif observation not in self.possible_observations:
+            raise ValueError('Observation {} not available'.format(observation))
+        else:
+            self.validation_data = self.get_validation_data(observation)
+
+        #plotting variables
+        self.ncolumns = int(ncolumns)
+        self._color_iterator = ('C{}'.format(i) for i in count())
+
+        #setup subplot configuration and get z cuts for each plot
+        self.z_lo, self.z_hi = self.init_plots(zlo, zhi, zint)
+        zmin = np.min(self.z_lo)
+        zmax = np.max(self.z_hi)
+        self.filters = [(lambda z: (z > zmin) & (z < zmax), self.zlabel)]
+
+        #setup summary plot
+        self.summary_fig, self.summary_ax = plt.subplots(self.nrows, self.ncolumns, sharex='col')
+        self.summary_fig.text(0.04, 0.5, self.yaxis, va='center', rotation='vertical') #setup a common axis label
+        #could plot summary validation data here if available but would need to evaluate labels, bin values etc.
+        #otherwise setup a check so that validation data is plotted only once on summary plot
+        self.first_pass = True
+
+        self._other_kwargs = kwargs
+
+
+    def init_plots(self, zlo, zhi, zint):
+        #get magnitude cuts based on validation data or default limits (only catalog data plotted)
+        if len(self.validation_data )> 0:
```

Comment:
change to `if  not self.validation_data:`

Reply:
Done